src/lib/storage/space_maker.rb

```diff
@@ -38,39 +37,133 @@ module Storage
+      # @param candidate_disks [Array<string>] device names of disks to install to
+      #
+      # @param linux_partitions [Array<string>] device names of existing Linux
+      #        partitions
       #
-      # @param disk_analyzer [DiskAnalyzer]
+      # @param windows_partitions [Array<string>] device names of Windows
+      #        partitions that can possibly be resized
       #
-      def initialize(volumes, settings, disk_analyzer)
-        @volumes  = volumes
-        @settings = settings
-        @disk_analyzer = disk_analyzer
+      # @param devicegraph [Storage::Devicegraph] devicegraph to use for any
+      #        changes, typically StorageManager.instance.devicegraph("proposal")
+      #
+      def initialize(settings:,
+                     volumes:            [],
+                     candidate_disks:    [],
+                     linux_partitions:   [],
+                     windows_partitions: [],
+                     devicegraph:        nil)
+        @settings           = settings
+        @volumes            = volumes
+        @candidate_disks    = candidate_disks
+        @linux_partitions   = linux_partitions
+        @windows_partitions = windows_partitions
+        @devicegraph        = devicegraph || StorageManager.instance.staging
+        @free_space         = []
       end
 
       # Try to detect empty (unpartitioned) space.
+      #
       def find_space
-        # TO DO
+        total_min_size     = total_vol_sizes(:min_size)
+        total_max_size     = total_vol_sizes(:max_size)
+        total_desired_size = total_desired_sizes
+        log.info("Total min     size:   #{total_min_size}")
+        log.info("Total desired size:   #{total_desired_size}")
+        log.info("Total max     size:   #{total_max_size}")
+        @free_space = []
+        @candidate_disks.each do |disk_name|
+          begin
+            log.info("Trying to find unpartitioned space on #{disk_name}")
+            disk = ::Storage::Disk.find(@devicegraph, disk_name)
+            disk.partition_table.unused_partition_slots.each do |slot|
+              size = DiskSize.new(slot.region.to_kb(slot.region.length))
+              log.info("Found slot: #{slot.region} size #{size} on #{disk_name}")
+              @free_space << slot
+            end
+          rescue RuntimeError => ex # FIXME: rescue ::Storage::Exception when SWIG bindings are fixed
+            log.info("CAUGHT exception #{ex}")
+            # FIXME: Handle completely empty disks (no partition table) as empty space
+          end
+        end
       end
 
       # Use force to create space: Try to resize an existing Windows
       # partition or delete partitions until there is enough free space.
+      #
       def make_space
         # TO DO
       end
 
       # Resize an existing MS Windows partition to free up disk space.
+      #
       def resize_windows_partition(partition)
         # TO DO
         partition
       end
+
+      # Delete all partitions on a disk.
```

Comment:
According to the code it seems that an empty partition table is created. This should be documented here.